server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/LocalDeliveryTest.java

```diff
@@ -44,83 +46,145 @@
 
+        LocalDelivery localDelivery = new LocalDelivery();
+        localDelivery.setDomainList(domainList);
+        localDelivery.setRrt(recipientRewriteTable);
+        localDelivery.setMailboxManager(mailboxManager);
+        localDelivery.setUsersRepository(usersRepository);
+
+        MailboxSession.User user = mock(MailboxSession.User.class);
+        MailboxSession session = mock(MailboxSession.class);
         when(session.getPathDelimiter()).thenReturn('.');
-        when(mailboxManager.createSystemSession(any(String.class), any(Logger.class))).thenReturn(session);
+        try {
+            when(mailboxManager.createSystemSession(any(String.class), any(Logger.class))).thenReturn(session);
+        } catch (MailboxException e) {
+            throw Throwables.propagate(e);
+        }
         when(session.getUser()).thenReturn(user);
+
+        return ImmutableList.of(
+            new Object[]{new Parameter(usersRepository, mailboxManager, sieveRepository, sieveLocalDelivery, user)},
+            new Object[]{new Parameter(usersRepository, mailboxManager, sieveRepository, localDelivery, user)}
+        );
     }
 
+    @Parameterized.Parameter
+    public Parameter parameter;
+
     @Test
     public void mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOn() throws Exception {
-        when(usersRepository.supportVirtualHosting()).thenReturn(true);
+        // Given
         String username = "receiver@domain.com";
-        when(sieveRepository.getActive(username)).thenThrow(new ScriptNotFoundException());
         MailboxPath inbox = new MailboxPath("#private", username, "INBOX");
         MessageManager messageManager = mock(MessageManager.class);
-        when(mailboxManager.getMailbox(eq(inbox), any(MailboxSession.class))).thenReturn(messageManager);
-        when(user.getUserName()).thenReturn(username);
 
-        Mail mail = createMail();
+        when(parameter.getUsersRepository().supportVirtualHosting()).thenReturn(true);
+        doThrow(new ScriptNotFoundException()).when(parameter.getSieveRepository()).getActive(username);
+        when(parameter.getMailboxManager().getMailbox(eq(inbox), any(MailboxSession.class))).thenReturn(messageManager);
+        when(parameter.getUser().getUserName()).thenReturn(username);
 
-        sieveLocalDelivery.init(new FakeMailetConfig());
-        sieveLocalDelivery.service(mail);
+        // When
+        Mail mail = createMail();
+        parameter.getMailet().init(new FakeMailetConfig());
+        parameter.getMailet().service(mail);
 
+        // Then
         verify(messageManager).appendMessage(any(InputStream.class), any(Date.class), any(MailboxSession.class), eq(true), any(Flags.class));
     }
 
     @Test
     public void mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOff() throws Exception {
-        when(usersRepository.supportVirtualHosting()).thenReturn(false);
+        // Given
```

Comment:
Yes, here testing logic is complicated. I prefer making explicit the different parts of the test. On simpler tests, it is not required, as blank lines or naming might do the job.